### ISSUE TITLE
Crash: java.lang.IllegalStateException: Unsupported concurrent change during composition. A state object was modified by composition as well as being modified outside composition

### ISSUE THREAD

**Issue by Caij** [coil-kt/coil#2352]
version: 3.0.0-alpha07

when ImageLoader set interceptor Dispatcher, There may be crash issues.

demo: https://github.com/Caij/CoilDemo

Open the app, wait for the image load finish, and click hello

```
at androidx.compose.runtime.Recomposer.applyAndCheck(Recomposer.kt:1235)
at androidx.compose.runtime.Recomposer.performRecompose(Recomposer.kt:1646)
at androidx.compose.runtime.Recomposer.access$performRecompose(Recomposer.kt:127)
at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:583)
at androidx.compose.runtime.Recomposer$runRecomposeAndApplyChanges$2$1.invoke(Recomposer.kt:551)
at androidx.compose.ui.platform.AndroidUiFrameClock$withFrameNanos$2$callback$1.doFrame(AndroidUiFrameClock.android.kt:41)
at androidx.compose.ui.platform.AndroidUiDispatcher.performFrameDispatch(AndroidUiDispatcher.android.kt:109)
at androidx.compose.ui.platform.AndroidUiDispatcher.access$performFrameDispatch(AndroidUiDispatcher.android.kt:41)
at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.doFrame(AndroidUiDispatcher.android.kt:69)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1449)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1459)
at android.view.Choreographer.doCallbacks(Choreographer.java:1089)
at android.view.Choreographer.doFrame(Choreographer.java:998)
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1431)
at android.os.Handler.handleCallback(Handler.java:942)
at android.os.Handler.dispatchMessage(Handler.java:99)
```

**Comment by colinrtwhite** [comment]
Thanks for the repro project. Tracking the issue [here](https://github.com/coil-kt/coil/issues/2346).

**Comment by Caij** [comment]
Hello. The demo provided in my issue will definitely crash.
This is a new issue caused by https://github.com/coil-kt/coil/pull/2205. I believe it shouldn't be in the 'not planned' status.

**Comment by Caij** [comment]
when ImageLoader set interceptor Dispatcher,  onSuccess is dispatched in my custom dispatcher, So it might cause an issue with multi-threaded state modification.
![image](https://github.com/coil-kt/coil/assets/8597839/98c66deb-7416-4b23-8ab5-0c076c419106)

**Comment by colinrtwhite** [comment]
@Caij I think this should be fixed in `3.0.0-alpha08`. `alpha07` used `Dispatchers.Unconfined`, which could result in that behaviour.

**Comment by Caij** [comment]
thanks, I have tested using the latest version and there is no problem
